src/supply_info.rs

```diff
@@ -0,0 +1,99 @@
+//! Structs and logic related to supply information management for assets.
+
+use std::collections::{hash_map, HashMap};
+
+use crate::{issuance::Error, note::AssetBase, value::ValueSum};
+
+/// Represents the amount of an asset and its finalization status.
+#[derive(Debug, Clone)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct AssetSupply {
+    /// The amount of the asset.
+    pub amount: ValueSum,
+    /// Whether or not the asset is finalized.
+    pub is_finalized: bool,
+}
+
+impl AssetSupply {
+    /// Creates a new AssetSupply instance with the given amount and finalization status.
+    pub fn new(amount: ValueSum, is_finalized: bool) -> Self {
+        Self {
+            amount,
+            is_finalized,
+        }
+    }
+}
+
+/// Contains information about the supply of assets.
+#[derive(Debug, Clone, Default)]
+pub struct SupplyInfo {
+    /// A map of asset bases to their respective supply information.
+    pub assets: HashMap<AssetBase, AssetSupply>,
+}
+
+impl SupplyInfo {
+    /// Inserts or updates an asset's supply information in the supply info map.
+    /// If the asset exists, adds the amounts (unconditionally) and updates the finalization status
+    /// (only if the new supply is finalized). If the asset is not found, inserts the new supply.
+    pub fn add_supply(&mut self, asset: AssetBase, new_supply: AssetSupply) -> Result<(), Error> {
+        match self.assets.entry(asset) {
+            hash_map::Entry::Occupied(mut entry) => {
+                let supply = entry.get_mut();
+                supply.amount =
+                    (supply.amount + new_supply.amount).ok_or(Error::ValueSumOverflow)?;
+                supply.is_finalized |= new_supply.is_finalized;
+            }
+            hash_map::Entry::Vacant(entry) => {
+                entry.insert(new_supply);
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::value::ValueSum;
+
+    #[test]
+    fn test_add_supply_valid() {
+        let mut supply_info = SupplyInfo::default();
```

Comment:
In this project, there are many custom types. Let's avoid `default()` and add an explicit `new()` constructor which is well defined in the scope of our new code.

Reply:
It's recommended by Clippy:

`
error: you should consider adding a `Default` implementation for `SupplyInfo`
  --> src/supply_info.rs:35:5
   |
35 | /     pub fn new() -> Self {
36 | |         Self {
37 | |             assets: HashMap::new(),
38 | |         }
39 | |     }
   | |_____^
   |
   = note: `-D clippy::new-without-default` implied by `-D warnings`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#new_without_default
help: try adding this
   |
34 + impl Default for SupplyInfo {
35 +     fn default() -> Self {
36 +         Self::new()
37 +     }
38 + }
   |
`

To avoid the warning but still use new, we can either disable the Clippy warning:

`
#[allow(clippy::new_without_default)]
pub fn new() -> Self {
    Self {
        assets: HashMap::new(),
    }
}
`

Or add a default implementation as it recommends:

`
impl Default for SupplyInfo {
    fn default() -> Self {
        Self::new()
    }
}
`

For now, I've fixed it by using the second option (added the default implementation that refers to `new`). And so I use `new` in the rest of code instead of `default`.